README.md
# task-timelog-hook
> A [taskwarrior](https://taskwarrior.org) hook that logs task start/stop times to a [timeclock file](https://hledger.org/hledger.html#timeclock-format), that [ledger](https://ledger-cli.org) and [hledger](https://hledger.org) can read. 

- _TLDR;_ If you enjoy the command-line and you want to track your time, including taskwarrior start/stop times, projects and tags, and you want to be able to easily read/ report/ query/ export your time-accounting data, then you should give this project a try. This approach is an alternative to [timewarrior](http://timewarrior.net). 
- _Warning_: experimenting with ledger and/or hledger can lead to vastly improved personal financial and time accounting! 

## INTRODUCTION

- [Taskwarrior](https://taskwarrior.org) is a command-line task-management program that is amazingly customizable, with projects and tags and reports and User Defined Attributes.. it's "the 800lb gorilla" in the CLI ToDo list jungle. Tasks can be started and stopped, and those times recorded as annotations, but taskwarrior doesn't have the built-in functionality to track or report on the those intervals. That's where ledger and hledger shine! 

- [ledger](https://ledger-cli.org) and [hledger](https://hledger.org) are both mature [command-line accounting programs](http://plaintextaccounting.org) with active development communities. They are work-alikes, and while [not identical](https://hledger.org/faq.html#how-is-hledger-different-from-ledger-) (ledger written in c++, hledger in haskell) they share most file-formats and many commands. One of the things they have in common is the ability to query and report from timeclock (a.k.a. timelog) formatted text files. The terms "timeclock" and "timelog" seem to be used interchangeably, and that's fine as long as the file extension is "*.timeclock". 

## INSTALL 
### Requirements
- Taskwarrior
- Taskwarrior hook
- ledger or hledger 
- 
## CONFIGURE
* files location
* editor

## RECOMMENDED
- timedot (for hledger users)
- timedot-vim
- t and t. scripts

## USAGE
### Timelog files
Timelog entries can be created taskwarrior start/stop commands, or directly by editing the *.timeclock file, or by using "t".
### t 
"t" is a handy script to work with (h)ledger and timelog files
```
Usage: t<space><action> or t<CR> for status        "t" is for "timelog"
  actions:
     i|in <account:sub> [desc] [-- comment]      td|today - balance today
     o|out [comment]                            yd|yesterday - bal yesterday
     a|accounts - list accounts used            yd^ - balance for 2 days ago
     b|bal - balance report [args]              tw|thisweek - bal for this week
  *  c|comm - add comment                       lw|lastweek - bal for last week
  *  d|dot - open timedot file (hledger)        tm|thismonth - bal for this mo
     e|edit - edit timelog file                 lm|lastmonth - bal for last mo
     f|file - show timelog file         _             
     g|grep - grep [args]              | |_     For report args and options see
     h|help - (you're looking at it)   | __|    ledger-cli.org or man ledger 
  *  l|log - record previous event     | |_      or  
     p|print - print [args]             \__|    hledger.org, run hledger<CR>
     r|reg - register [args]                   
     s|stats                                    For user configs edit this file
     t|tail - show end of timelog               For corrections edit timelog 
  *  u|ui - open in hledger-ui                  For more details see README.md
  *  v|version                                  
  *  w|web - open timelog in browser
  *  z|zip - backup files                       Please report issues/fixes to 
      ( * = planned )          https://github.com/linuxcaffe/task-timelog-hook/
```
### Timedot

## CREDITS

This hook is based entirely on https://gist.github.com/wbsch/d977b0ac29aa1dfa4437 by Wilhelm Schürmann in 2015, and it works just great! I'm expanding the gist to a github repo to provide documentation and tweaks where needed.

## COMMENTS
## CONTRIBUTIONS
Bug reports, suggestions, pull requests welcome https://github.com/linuxcaffe/task-timelog-hook/issues



CHANGES.md
# CHANGES.md


Sun Nov 21 16:28:50 EST 2021
working on t, 
- changed it from sh to bash
- echo action on in or out
- added "i" and "o"
- allowed following arguments (payee field) 2space separated
- ignore all lines that don't start with "i " or "o "
- reworked use-last-account, if none specified
- prettied prompts

Tue Nov 23 16:58:37 EST 2021
- added timezone variable
- rewrote status function, bind to "T"<no args>


on-modify-timelog.py
#!/usr/bin/env python
#
# Writes task start/stop times to a timelog formatted file.
# You might need to adjust LEDGERFILE, or set the TIMELOG environment variable.
#
# Example reports, after using start/stop on a task:
# ledger -f /path/to/timelog.ledger print
# ledger -f /path/to/timelog.ledger register
#
# Projects, tags, and UUIDs are fully supported and queryable from ledger.
#
#
# 2015-03-05 wbsch
#   - Now with "I forgot to start/stop this task!" convenience features:
#       "task $id start $x"
#       "task $id stop $x"
#       "task $id done $x"
#     Where $x is the time in minutes you want the entry in your timelog
#     file to be backdated. Note that this is not properly displayed in
#     Taskwarrior itself, but only in your timelog file.
#
#     Note: This will only work on Taskwarrior 2.4.2+ due to a bug in
#           earlier versions. The basic time tracking functionality will
#           work on 2.4.1+.
#
#
# May the Holy Python forgive me for this mess.
#

import calendar
import json
import os
import re
import sys
from datetime import datetime
from datetime import timedelta


LEDGERFILE = "%s/.task/hooks/task-timelog-hook/tw.timelog" % os.getenv('HOME')

if 'TIMELOG' in os.environ:
    LEDGERFILE = os.environ['TIMELOG']


def adjust_date(d, adjust_by):
    if not isinstance(d, datetime):
        d = tw_to_dt(d)
    d -= timedelta(minutes=int(adjust_by))
    return d

def tw_to_dt(s):
    """ Taskwarrior JSON date ---> datetime object. """
    return datetime.strptime(s, "%Y%m%dT%H%M%SZ")

def dt_to_tw(d):
    """ datetime object ---> Taskwarrior JSON date. """
    return d.strftime("%Y%m%dT%H%M%SZ")


old = json.loads(sys.stdin.readline())
new = json.loads(sys.stdin.readline())

annotation_added = ('annotations' in new and not 'annotations' in old) \
                    or \
                 ('annotations' in new and 'annotations' in old and \
                  len(new['annotations']) > len(old['annotations']))


# task started
if ('start' in new and not 'start' in old) and annotation_added:
    new['annotations'].sort(key=lambda anno: anno['entry'])
    m = re.match('^[0-9]+$', new['annotations'][-1]['description'])
    if m:
        new['start'] = dt_to_tw(adjust_date(new['start'], int(m.group(0))))
        new['annotations'] = new['annotations'][:-1]
        if not new['annotations']:
            del new['annotations']
        print("Timelog: Started task %s minutes ago." % m.group(0))
        
        if tw_to_dt(new['start']) < tw_to_dt(new['entry']):
            new['entry'] = new['start']

# task stopped
if 'start' in old and not 'start' in new:
    started_utc = tw_to_dt(old['start'])
    started_ts = calendar.timegm(started_utc.timetuple())
    started = datetime.fromtimestamp(started_ts)
    stopped = datetime.now()

    if annotation_added:
        new['annotations'].sort(key=lambda anno: anno['entry'])
        m = re.match('^[0-9]+$', new['annotations'][-1]['description'])
        if m:
            new['annotations'] = new['annotations'][:-1]
            if not new['annotations']:
                del new['annotations']
            stopped = adjust_date(stopped, m.group(0))
            if stopped < started:
                print("ERROR: Stop date -%s minutes would be before the start date!" % m.group(0))
                sys.exit(1)
            print("Timelog: Stopped task %s minutes ago." % m.group(0))
    
    entry = "i " + started.strftime("%Y/%m/%d %H:%M:%S")
    entry += " "
    entry += new['project'].replace('.', ':') if 'project' in new else "no project"
    entry += "  " + new['description'] + "\n"
    entry += "o " + stopped.strftime("%Y/%m/%d %H:%M:%S")
    entry += "  ;"
    entry += " :" + ":".join(new['tags']) + ":" if 'tags' in new else ""
    entry += " uuid: " + new['uuid']
    entry += "\n\n"
    with open(LEDGERFILE, "a") as ledger:
        ledger.write(entry.encode("utf-8"))

print(json.dumps(new))


on-modify-timeclock.py
#!/usr/bin/env python
#
# Writes task start/stop times to a timelog formatted file.
# You might need to adjust LEDGERFILE, or set the TIMELOG environment variable.
#
# Example reports, after using start/stop on a task:
# ledger -f /path/to/timelog.ledger print
# ledger -f /path/to/timelog.ledger register
#
# Projects, tags, and UUIDs are fully supported and queryable from ledger.
#
#
# 2015-03-05 wbsch
#   - Now with "I forgot to start/stop this task!" convenience features:
#       "task $id start $x"
#       "task $id stop $x"
#       "task $id done $x"
#     Where $x is the time in minutes you want the entry in your timelog
#     file to be backdated. Note that this is not properly displayed in
#     Taskwarrior itself, but only in your timelog file.
#
#     Note: This will only work on Taskwarrior 2.4.2+ due to a bug in
#           earlier versions. The basic time tracking functionality will
#           work on 2.4.1+.
#
#
# May the Holy Python forgive me for this mess.
#

import calendar
import json
import os
import re
import sys
from datetime import datetime
from datetime import timedelta


LEDGERFILE = "%s/.task/hooks/task-timelog-hook/tw.timeclock" % os.getenv('HOME')

if 'TIMELOG' in os.environ:
    LEDGERFILE = os.environ['TIMELOG']


def adjust_date(d, adjust_by):
    if not isinstance(d, datetime):
        d = tw_to_dt(d)
    d -= timedelta(minutes=int(adjust_by))
    return d

def tw_to_dt(s):
    """ Taskwarrior JSON date ---> datetime object. """
    return datetime.strptime(s, "%Y%m%dT%H%M%SZ")

def dt_to_tw(d):
    """ datetime object ---> Taskwarrior JSON date. """
    return d.strftime("%Y%m%dT%H%M%SZ")


old = json.loads(sys.stdin.readline())
new = json.loads(sys.stdin.readline())

annotation_added = ('annotations' in new and not 'annotations' in old) \
                    or \
                 ('annotations' in new and 'annotations' in old and \
                  len(new['annotations']) > len(old['annotations']))


# task started
if ('start' in new and not 'start' in old) and annotation_added:
    new['annotations'].sort(key=lambda anno: anno['entry'])
    m = re.match('^[0-9]+$', new['annotations'][-1]['description'])
    if m:
        new['start'] = dt_to_tw(adjust_date(new['start'], int(m.group(0))))
        new['annotations'] = new['annotations'][:-1]
        if not new['annotations']:
            del new['annotations']
        print("Timelog: Started task %s minutes ago." % m.group(0))
        
        if tw_to_dt(new['start']) < tw_to_dt(new['entry']):
            new['entry'] = new['start']

# task stopped
if 'start' in old and not 'start' in new:
    started_utc = tw_to_dt(old['start'])
    started_ts = calendar.timegm(started_utc.timetuple())
    started = datetime.fromtimestamp(started_ts)
    stopped = datetime.now()

    if annotation_added:
        new['annotations'].sort(key=lambda anno: anno['entry'])
        m = re.match('^[0-9]+$', new['annotations'][-1]['description'])
        if m:
            new['annotations'] = new['annotations'][:-1]
            if not new['annotations']:
                del new['annotations']
            stopped = adjust_date(stopped, m.group(0))
            if stopped < started:
                print("ERROR: Stop date -%s minutes would be before the start date!" % m.group(0))
                sys.exit(1)
            print("Timelog: Stopped task %s minutes ago." % m.group(0))
    
    entry = "i " + started.strftime("%Y/%m/%d %H:%M:%S")
    entry += " "
    entry += new['project'].replace('.', ':') if 'project' in new else "no project"
    entry += "  " + new['description'] + "\n"
    entry += "o " + stopped.strftime("%Y/%m/%d %H:%M:%S")
    entry += "  ; "
# TODO: separate tags (is now   tag1,tag2,tag3:,   should be   tag1:, tag2:, tag3:,)
    entry += "" .join(new['tags']) + ":," if 'tags' in new else ""
# TODO: extract uuid.short, not (default) uuid.long
    entry += " uuid: " + new['uuid']
    entry += "\n\n"
    with open(LEDGERFILE, "a") as ledger:
        ledger.write(entry.encode("utf-8"))

print(json.dumps(new))
